pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,455 @@
+// bookmark to an event record returned from GetEvents() when it makes sense for the user.
+// https://learn.microsoft.com/en-us/windows/win32/wes/bookmarking-events
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithStartAfterBookmark(bookmark evtbookmark.Bookmark) PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.SetBookmark(bookmark)
+	}
+}
+
+// WithStartAtOldestRecord will start the subscription from the oldest record in the event log.
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithStartAtOldestRecord() PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.subscribeOriginFlag = evtapi.EvtSubscribeStartAtOldestRecord
+	}
+}
+
+// WithSubscribeFlags can be used to manually set EVT_SUBSCRIBE_FLAGS
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithSubscribeFlags(flags uint) PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.subscribeFlags = flags
+	}
+}
+
+// WithNotifyEventsAvailable enables the EventsAvailable() channel
+func WithNotifyEventsAvailable() PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.useNotifyChannel = true
+	}
+}
+
+func (q *pullSubscription) EventsAvailable() <-chan struct{} {
+	return q.notifyEventsAvailable
+}
+
+func (q *pullSubscription) Running() bool {
+	return q.started
+}
+
+func (q *pullSubscription) SetBookmark(bookmark evtbookmark.Bookmark) {
+	q.bookmark = bookmark
+	q.subscribeOriginFlag = evtapi.EvtSubscribeStartAfterBookmark
+}
+
+func (q *pullSubscription) Start() error {
+
+	if q.started {
+		return fmt.Errorf("Query subscription is already started")
+	}
+
+	var bookmarkHandle evtapi.EventBookmarkHandle
+	if q.bookmark != nil {
+		bookmarkHandle = q.bookmark.Handle()
+	}
+
+	// create subscription
+	hSubWait, err := newSubscriptionWaitEvent()
+	if err != nil {
+		return err
+	}
+	hSub, err := q.eventLogAPI.EvtSubscribe(
+		hSubWait,
+		q.channelPath,
+		q.query,
+		bookmarkHandle,
+		q.subscribeOriginFlag|q.subscribeFlags)
+	if err != nil {
+		safeCloseNullHandle(windows.Handle(hSubWait))
+		return err
+	}
+
+	// alloc reusable storage for EvtNext output
+	q.evtNextStorage = make([]evtapi.EventRecordHandle, q.eventBatchCount)
+
+	q.waitEventHandle = hSubWait
+	q.subscriptionHandle = hSub
+
+	if q.useNotifyChannel {
+		hStopWait, err := newStopWaitEvent()
+		if err != nil {
+			evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+			safeCloseNullHandle(windows.Handle(hSubWait))
+			return err
+		}
+		q.stopEventHandle = hStopWait
+
+		// Query loop management
+		q.notifyStop = make(chan struct{})
+		q.notifyNoMoreItems = make(chan struct{})
+		q.notifyNoMoreItemsComplete = make(chan struct{})
+		q.notifyEventsAvailable = make(chan struct{})
+		q.notifyEventsAvailableLoopDone = make(chan struct{})
```

Comment:
oh I 100% agree. I agonized over translating the windows signal event into a go channel and hit a lot of bugs along the way, each bug adding a new channel...

I think it can be made simpler if we accept that it is okay to notify the user that events are available when there actually aren't any. which does happen anyway on the first call because the event starts set.
```
// EvtNext called, there are no more items to read, this case
// allows us to cancel sending notifyEventsAvailable to the user.
```

Reply:
refactored in https://github.com/DataDog/datadog-agent/pull/19761/commits/3a7c80ba795c513cf2c3349fdf69461fd7860322